### PR TITLE
1.x: distinctUntilChanged with direct value comparator - alternative

### PR DESCRIPTION
**Pull request by akarnokd**
Alternative implementation to #4001.

### CODE REVIEW

File: src/main/java/rx/internal/operators/OperatorDistinctUntilChanged.java [hunk 1]
```diff
@@ -68,7 +83,16 @@ public void onNext(T t) {
...
+                    Boolean comparison;
+                    
+                    try {
+                        comparison = comparator.call(currentKey, key);
+                    } catch (Throwable e) {
+                        Exceptions.throwOrReport(e, child, key);
+                        return;
+                    }
+                    
+                    if (!comparison) {
```
Comment (JakeWharton) [inline src/main/java/rx/internal/operators/OperatorDistinctUntilChanged.java:95]: There's a subtle NPE potential here. Not sure if you want to explicitly guard against it from a misbehaving `comparator` implementation.

File: src/main/java/rx/internal/operators/OperatorDistinctUntilChanged.java [hunk 2]
```diff
@@ -48,6 +50,19 @@
 
     public OperatorDistinctUntilChanged(Func1<? super T, ? extends U> keySelector) {
         this.keySelector = keySelector;
+        this.comparator = this;
```
Comment (JakeWharton) [inline src/main/java/rx/internal/operators/OperatorDistinctUntilChanged.java:53]: Cute trick

File: src/main/java/rx/internal/operators/OperatorDistinctUntilChanged.java (hunk 1 above)
Comment (akarnokd) [inline src/main/java/rx/internal/operators/OperatorDistinctUntilChanged.java:95]: Good catch! I'll change the comparison to be primitive boolean to trigger NPE inside the try-block if necessary.

**Comment by JakeWharton** [comment]
lgtm :+1:

File: src/main/java/rx/internal/operators/OperatorDistinctUntilChanged.java (hunk 1 above)
Comment (akarnokd) [inline src/main/java/rx/internal/operators/OperatorDistinctUntilChanged.java:95]: Done.

**Comment by akarnokd** [comment]
Thanks Jake for the review!

**Comment by stevegury** [comment]
👍

**Pull request merged**